src/main/java/com/artipie/api/artifactory/AddUpdateUserSlice.java

```diff
@@ -23,30 +23,97 @@
+import javax.json.JsonObject;
+import javax.json.JsonReader;
 import org.reactivestreams.Publisher;
 
 /**
  * Artifactory `PUSH/PUT /api/security/users/{userName}` endpoint,
- * adds/updates user record in credentials.
+ * updates/adds user record in credentials.
+ *
  * @since 0.10
- * @todo #444:30min Implement this slice to add/update user from credentials by user name
- *  obtained from request line, path format is `/api/security/users/{userName}`. Password should
- *  be obtain from `password` field from json request body.
- *  Use Credentials#add(java.lang.String, java.lang.String) method to perform the operation and
- *  return 200 OK status. Do not forget to test this class and add it to ArtipieApi,
- *  check GetUserSlice as an example.
  */
 public final class AddUpdateUserSlice implements Slice {
 
+    /**
+     * Artipie settings.
+     */
+    private final Settings settings;
+
+    /**
+     * Ctor.
+     *
+     * @param settings Artipie setting
+     */
+    public AddUpdateUserSlice(final Settings settings) {
+        this.settings = settings;
+    }
+
     @Override
     public Response response(final String line, final Iterable<Map.Entry<String, String>> headers,
         final Publisher<ByteBuffer> body) {
-        throw new NotImplementedException("Not implemented yet");
+        final Response res;
+        final Matcher matcher = GetUserSlice.PTRN.matcher(
+            new RequestLineFrom(line).uri().toString()
+        );
+        if (matcher.matches()) {
+            final CompletionStage<Optional<String>> pswd = AddUpdateUserSlice.password(body);
+            final String username = matcher.group("username");
+            res = new AsyncResponse(
+                pswd.thenCompose(
+                    passw -> passw.map(
+                        haspassw -> this.settings.credentials()
+                            .thenCompose(
+                                cred -> cred.add(username, haspassw)
+                                    .thenApply(ok -> new RsWithStatus(RsStatus.OK)))
+                ).orElse(CompletableFuture.completedFuture(new RsWithStatus(RsStatus.NOT_FOUND))))
+            );
+        } else {
+            res = new RsWithStatus(RsStatus.BAD_REQUEST);
+        }
+        return res;
     }
 
+    /**
+     * Extracts password string from the response body.
+     *
+     * @param body Response body
+     * @return Password string as completion.
+     */
+    private static CompletionStage<Optional<String>> password(final Publisher<ByteBuffer> body) {
+        return new PublisherAs(body)
+            .bytes().thenApply(
+                bytes -> {
+                    final Optional<String> pswd;
+                    if (bytes.length == 0) {
+                        pswd = Optional.empty();
+                    } else {
+                        final JsonObject root;
+                        try (JsonReader reader =
+                            Json.createReader(new ByteArrayInputStream(bytes))
```

Comment:
@genryxy there is no need in using try with resources here as `JsonReader` closes underlying input source on `close()` and `ByteArrayInputStream` does nothing on `close()`.